src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/DiagnosticOr.cs

```diff
@@ -32,33 +32,67 @@ public abstract record DiagnosticOr<T>
+            public override bool HasValue => false;
+            public override bool HasDiagnostic => true;
             public override T Value => throw new InvalidOperationException();
-            public override DiagnosticInfo Diagnostic => _diagnostic;
+            public override ImmutableArray<DiagnosticInfo> Diagnostics => _diagnostics.Array;
         }
 
         private sealed record Val : DiagnosticOr<T>
         {
             private readonly T _value;
             internal Val(T value) => _value = value;
-            public override bool IsValue => true;
+            public override bool HasValue => true;
+            public override bool HasDiagnostic => false;
             public override T Value => _value;
-            public override DiagnosticInfo Diagnostic => throw new InvalidOperationException();
+            public override ImmutableArray<DiagnosticInfo> Diagnostics => throw new InvalidOperationException();
+        }
+
+        private sealed record ValueAndDiagnostic : DiagnosticOr<T>
+        {
+            private readonly T _value;
+            private readonly SequenceEqualImmutableArray<DiagnosticInfo> _diagnostics;
+            internal ValueAndDiagnostic(T value, ImmutableArray<DiagnosticInfo> diagnostics) => (_value, _diagnostics) = (value, diagnostics.ToSequenceEqual());
+            public override bool HasValue => true;
+            public override bool HasDiagnostic => true;
+            public override T Value => _value;
+            public override ImmutableArray<DiagnosticInfo> Diagnostics => _diagnostics.Array;
         }
 
+        /// <summary>
+        /// Adds a diagnostic to the <see cref="DiagnosticOr{T}.Diagnostics"/> property
+        /// </summary>
+        public DiagnosticOr<T> AddDiagnostic(DiagnosticInfo diagnostic) => this switch
+        {
+            Diag d => new Diag(d.Diagnostics.Add(diagnostic)),
+            Val v => new ValueAndDiagnostic(v.Value, ImmutableArray.Create(diagnostic)),
+            ValueAndDiagnostic vad => new ValueAndDiagnostic(vad.Value, vad.Diagnostics.Add(diagnostic)),
+            _ => throw new UnreachableException()
+        };
+
+        /// <summary>
+        /// Creates a new <see cref="DiagnosticOr{T}"/> with the <see cref="DiagnosticOr{T}.Value"/> set to <paramref name="value"/>
+        /// </summary>
+        public DiagnosticOr<T> WithValue(T value) => this switch
+        {
+            Diag d => new ValueAndDiagnostic(value, d.Diagnostics),
+            Val => new Val(value),
+            ValueAndDiagnostic vad => new ValueAndDiagnostic(value, vad.Diagnostics),
+            _ => throw new UnreachableException()
+        };
+
         /// <summary>
         /// Create a Diagnostic variant
         /// </summary>
-        public static DiagnosticOr<T> From(DiagnosticInfo diagnostic)
+        public static DiagnosticOr<T> From(params DiagnosticInfo[] diagnostic)
```

Comment:
The `params` pattern is rather expensive. Can we have a single _and_ a `params` overload? Is there one that is the more common case?

Reply:
Yeah, we actually never use more than one, I'll remove params